Add Adot workshop (#3537)
aws/eks-anywhere: docs/content/en/docs/workshops/packages/adot/adot_amp_amg.md

---
title: "ADOT with AMP and AMG"
linkTitle: "ADOT with AMP and AMG"
weight: 2
date: 2022-09-30
description: >  
---

> This tutorial demonstrates how to config ADOT package to scrap metrics from an EKS Anywhere (EKS-A) cluster, and send them to [Amazon Managed Service for Prometheus](https://aws.amazon.com/prometheus/) (AMP) and [Amazon Managed Grafana](https://aws.amazon.com/grafana/) (AMG).

Author: Updated.

This tutorial walks through the following procedures:
- [Create an AMP workspace](#create-an-amp-workspace);

> - [Create an EKS-A cluster with IAM Roles for Service Account (IRSA)](#create-an-eks-a-cluster-with-irsa);

Author: Acknowledged. I updated it based on your suggestion.

> - [Install EKS-A ADOT package](#install-eks-a-adot-package);

Author: Updated

- [Create an AMG workspace and connect to the AMP workspace](#create-an-amg-workspace-and-connect-to-the-amp-workspace).

{{% alert title="Note" color="primary" %}}

- We included `Test` sections below for critical steps to help users to validate they have completed such procedure properly. We recommend going through them in sequence as checkpoints of the progress.
- We recommend creating all resources in the `us-west-2` region.

{{% /alert %}}

## Create an AMP workspace

An AMP workspace is created to receive metrics from the ADOT package, and respond to query requests from AMG. Follow steps below to complete the set up:

1. Open the AMP console at [https://console.aws.amazon.com/prometheus/.](https://console.aws.amazon.com/prometheus/)

1. Choose region `us-west-2` from the top right corner.

1. Click on `Create` to create a workspace.

> 1. Type an workspace alias (`adot-amp-test` as an example), and click on `Create workspace`.

Author: Updated

    ![ADOT AMP Create Workspace](/images/adot_amp_create_ws.png)

1. Make notes of the URLs displayed for `Endpoint - remote write URL` and `Endpoint - query URL`. You'll need them when you configure your ADOT package to remote write metrics to this workspace and when you query metrics from this workspace. Make sure the workspace's `Status` shows `Active` before proceeding to the next step.

    ![ADOT AMP Identify URLs](/images/adot_amp_identify_urls.png)

For additional options (i.e. through CLI) and configurations (i.e. add a tag) to create an AMP workspace, refer to [AWS AMP create a workspace guide.](https://docs.aws.amazon.com/prometheus/latest/userguide/AMP-onboard-create-workspace.html)

> ## Create an EKS-A cluster with IRSA

Author: Updated

> To enable ADOT pods that run in EKS-A clusters to authenticate with AWS services, a user needs to set up IRSA at cluster creation. [EKS-A cluster spec for Pod IAM](https://anywhere.eks.amazonaws.com/docs/reference/clusterspec/optional/irsa/) gives step-by-step guidance on how to do so. There are a few things to keep in mind while working through the guide:

Reviewer: ```suggestion
To enable ADOT pods that run in EKS Anywhere clusters to authenticate with AWS services, a user needs to set up IRSA at cluster creation. [EKS Anywhere cluster spec for Pod IAM]({{< relref "../../../reference//clusterspec/optional/irsa/" >}}) gives step-by-step guidance on how to do so. There are a few things to keep in mind while working through the guide:
```

Author: Updated

    - add an annotation `eks.amazonaws.com/role-arn: <role-arn>` to the created service account.

    By default, the service account is installed in the `default` namespace with name `pod-identity-webhook`, and the annotation `eks.amazonaws.com/role-arn: <role-arn>` is not added automatically.

### IRSA Set Up Test
To ensure IRSA is set up properly in the cluster, a user can create an `awscli` pod for testing.
1. Apply the following yaml file in the cluster:
    ```bash
    kubectl apply -f - <<EOF
    apiVersion: v1
    kind: Pod
    metadata:
      name: awscli
    spec:
      serviceAccountName: pod-identity-webhook
      containers:
      - image: amazon/aws-cli
        command:
          - sleep
          - "infinity"
        name: awscli
        resources: {}
      dnsPolicy: ClusterFirst
      restartPolicy: Always
    EOF
    ```

1. Exec into the pod:
    ```
    kubectl exec -it awscli -- /bin/bash
    ```

1. Check if the pod can list AMP workspaces:
    ```
    aws amp list-workspaces --region=us-west-2
    ```

1. If the pod has issues listing AMP workspaces, re-visit IRSA set up guidance before proceeding to the next step.

1. Exit the pod:
    ```
    exit
    ```

> ## Install EKS-A ADOT package

Reviewer: ```suggestion
## Install the ADOT package
```

Author: Updated

> The EKS-A ADOT package will be created with three components:

Reviewer: ```suggestion
The EKS Anywhere ADOT package will be created with three components:
```

Author: Removed the cluster reference

1. the Prometheus Receiver, which is designed to be a drop-in replacement for a Prometheus Server and is capable of scraping metrics from microservices instrumented with the [Prometheus client library](https://prometheus.io/docs/instrumenting/clientlibs/);

1. the Prometheus Remote Write Exporter, which employs the remote write features and send metrics to AMP for long term storage;

1. the Sigv4 Authentication Extension, which enables ADOT pods to authenticate to AWS services.

Follow steps below to complete the ADOT package installation:

> 1. Update the following config file. Review comments carefully and replace everything that is wrapped with a `<>` tag. Note this configuration aims to mimic the Prometheus community helm chart. A user can tailor the scrap targets further by modifying the receiver section below. Refer to [ADOT package spec]({{< relref "../../../reference/packagespec/adot" >}}) for additional explanations of each section.

Author: The ../../../reference/packagespec/adot link does exist. I verified it by a local build. And I can confirm this is the link I want, not adot/_index.md.

> 1. Update the following config file. Review comments carefully and replace everything that is wrapped with a `<>` tag. Note this configuration aims to mimic the Prometheus community helm chart. A user can tailor the scrap targets further by modifying the receiver section below. Refer to [ADOT package spec]({{< relref "../../../reference/packagespec/adot" >}}) for additional explanations of each section.

Reviewer: Sorry, you are right. I meant that I thought the file should be _index.md, not the link. I probably need to do a rebase to get those file.

    <details>
      <summary>Click to expand ADOT package config</summary>
      
      ```yaml
      apiVersion: packages.eks.amazonaws.com/v1alpha1
      kind: Package
      metadata:
        name: my-adot
        namespace: eksa-packages
      spec:
        packageName: adot
        targetNamespace: default # this needs to match the namespace of the serviceAccount below
        config: |
          mode: deployment

          serviceAccount:
            # Specifies whether a service account should be created
            create: false
            # Annotations to add to the service account
            annotations: {}
            # Specifies the serviceAccount annotated with eks.amazonaws.com/role-arn.

>             name: "pod-identity-webhook" # name of the service account created at step Create an EKS-A cluster with IRSA

Reviewer: Change "Create an EKS-A cluster with IRSA" to however you modify the heading.

Author: Updated

                      - __meta_kubernetes_namespace
                      target_label: namespace
                    - action: replace
                      source_labels:
                      - __meta_kubernetes_pod_name
                      target_label: pod
                    - action: drop
                      regex: Pending|Succeeded|Failed|Completed
                      source_labels:
                      - __meta_kubernetes_pod_phase

            processors:
              batch/metrics:
                timeout: 60s

            exporters:
              logging:
                logLevel: info
              prometheusremotewrite:
                endpoint: "<AMP-WORKSPACE>/api/v1/remote_write" # Replace with your AMP workspace
                auth:
                  authenticator: sigv4auth
            
            service:
              extensions:
                - health_check
                - memory_ballast
                - sigv4auth
              pipelines:
                metrics:
                  receivers: [prometheus]
                  processors: [batch/metrics]
                  exporters: [logging, prometheusremotewrite]

      ```
    </details>

> 1. Bind additional roles to the service account `pod-identity-webhook` (created at step [Create an EKS-A cluster with IRSA](#create-an-eks-a-cluster-with-irsa)) by applying the following file in the cluster (using `kubectl apply -f <file-name>`). This is because `pod-identity-webhook` by design does not have sufficient permissions to scrap all Kubernetes targets listed in the ADOT config file above. If modifications are made to the Prometheus Receiver, make updates to the file below to add / remove additional permissions before applying the file.

Reviewer: ```suggestion
1. Bind additional roles to the service account `pod-identity-webhook` (created at step [Create an EKS-A cluster with IRSA](#create-an-eks-a-cluster-with-irsa)) by applying the following file in the cluster (using `kubectl apply -f <file-name>`). This is because `pod-identity-webhook` by design does not have sufficient permissions to scrape all Kubernetes targets listed in the ADOT config file above. If modifications are made to the Prometheus Receiver, make updates to the file below to add / remove additional permissions before applying the file.
```

Author: Updated

    <details>
      <summary>Click to expand clusterrole and clusterrolebinding config</summary>

      ```yaml
      ---
      apiVersion: rbac.authorization.k8s.io/v1
      kind: ClusterRole
      metadata:
        name: otel-prometheus-role
      rules:
        - apiGroups:
            - ""
          resources:
            - nodes
            - nodes/proxy
            - services
            - endpoints
            - pods
          verbs:
            - get
            - list
            - watch
        - apiGroups:
            - extensions
          resources:
            - ingresses
          verbs:
            - get
            - list
            - watch
        - nonResourceURLs:
            - /metrics
          verbs:
            - get

      ---
      apiVersion: rbac.authorization.k8s.io/v1
      kind: ClusterRoleBinding
      metadata:
        name: otel-prometheus-role-binding
      roleRef:
        apiGroup: rbac.authorization.k8s.io
        kind: ClusterRole
        name: otel-prometheus-role
      subjects:
        - kind: ServiceAccount
          name: pod-identity-webhook  # replace with name of the service account created at step Create an EKS-A cluster with IRSA

>           namespace: default  # replace with namespace where the service account was created at step Create an EKS-A cluster with IRSA

Reviewer: Not critical, but consider renaming the reference "Create an EKS-A cluster with IRSA" if you rename the heading.

Author: Done

2022-09-30T23:22:59.188Z	info	kubernetes/kubernetes.go:326	Using pod service account via in-cluster config	{"kind": "receiver", "name": "prometheus", "pipeline": "metrics", "discovery": "kubernetes"}
2022-09-30T23:22:59.188Z	info	kubernetes/kubernetes.go:326	Using pod service account via in-cluster config	{"kind": "receiver", "name": "prometheus", "pipeline": "metrics", "discovery": "kubernetes"}
2022-09-30T23:22:59.188Z	info	kubernetes/kubernetes.go:326	Using pod service account via in-cluster config	{"kind": "receiver", "name": "prometheus", "pipeline": "metrics", "discovery": "kubernetes"}
2022-09-30T23:22:59.189Z	info	pipelines/pipelines.go:106	Receiver started.	{"kind": "receiver", "name": "prometheus", "pipeline": "metrics"}
2022-09-30T23:22:59.189Z	info	healthcheck/handler.go:129	Health Check state change	{"kind": "extension", "name": "health_check", "status": "ready"}
2022-09-30T23:22:59.189Z	info	service/collector.go:215	Starting aws-otel-collector...	{"Version": "v0.21.1", "NumCPU": 2}
2022-09-30T23:22:59.189Z	info	service/collector.go:128	Everything is ready. Begin running and processing data.
...
```

> #### Check AMP end point using awscurl

Reviewer: ```suggestion
#### Check AMP endpoint using awscurl
```

Author: Done
